ext/scheduler/airflow/bucket/ossblob/ossblob.go

```diff
@@ -0,0 +1,405 @@
+package ossblob
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/aliyun/alibabacloud-oss-go-sdk-v2/oss"
+	"gocloud.dev/blob"
+	"gocloud.dev/blob/driver"
+	"gocloud.dev/gcerrors"
+)
+
+const (
+	// DefaultMaxKeys is the default maximum number of keys to retrieve in a List call.
+	defaultMaxKeys = 1000
+
+	// DefaultRangeBehavior is the default range behavior for OSS.
+	defaultRangeBehavior string = "standard"
+)
+
+// ossReader is an implementation of driver.Reader for OSS.
+// This reader is used in OSS RangeReader to read content of a blob in OSS.
+type ossReader struct {
+	body io.ReadCloser
+	raw  *oss.RangeReader
+}
+
+// NewOSSReader creates a new ossReader.
+func NewOSSReader(body io.ReadCloser, raw *oss.RangeReader) *ossReader {
+	return &ossReader{
+		body: body,
+		raw:  raw,
+	}
+}
+
+func (r *ossReader) Read(p []byte) (int, error) {
+	return r.body.Read(p)
+}
+
+func (r *ossReader) Close() error {
+	return r.body.Close()
+}
+
+func (r *ossReader) As(i interface{}) bool {
+	p, ok := i.(**oss.RangeReader)
+	if !ok {
+		return false
+	}
+	*p = r.raw
+	return true
+}
+
+// Attributes implements driver.Reader.Attributes.
+// For now this will return no attributes
+func (*ossReader) Attributes() *driver.ReaderAttributes {
+	return &driver.ReaderAttributes{}
+}
+
+// ossWriter is an implementation of driver.Writer for OSS.
+// pipereader & pipewriter is used so that a stream-like write can be done using
+// OSS PutObject API.
+type ossWriter struct {
+	req    oss.PutObjectRequest
+	pr     *io.PipeReader
+	pw     *io.PipeWriter
+	doneCh chan struct{}
+	err    error
+}
+
+// NewOSSWriter creates a new ossWriter.
+func NewOSSWriter(req oss.PutObjectRequest, pr *io.PipeReader, pw *io.PipeWriter) *ossWriter {
+	return &ossWriter{
+		req:    req,
+		pr:     pr,
+		pw:     pw,
+		doneCh: make(chan struct{}),
+	}
+}
+
+func (w *ossWriter) Write(p []byte) (int, error) {
+	return w.pw.Write(p)
+}
+
+func (w *ossWriter) Close() error {
+	w.pw.Close()
+
+	<-w.doneCh
+	return w.err
+}
+
+func (w *ossWriter) As(i interface{}) bool {
+	p, ok := i.(**oss.PutObjectRequest)
+	if !ok {
+		return false
+	}
+	*p = &w.req
+	return true
+}
+
+type ossBucket struct {
+	client *oss.Client
+	bucket string
+}
+
+// NewOSSBucket creates a new ossBucket.
+func NewOSSBucket(client *oss.Client, bucket string) *ossBucket {
```

Comment:
although this is very good for learning `ossblob`, we eventually need some extra effort to maintain this implementation, this is the thing that we need to avoid. (either we can contribute to gocloud repo, or ask ali to implement this, or omit this entirely)

Reply:
our interface `airflow.Bucket` has some methods that coupled with gocloud interface, I understand that's the reason why we implement blob here. instead of implementing blob, why don't we refactor `airflow.Bucket` interface so that it's not contain gocloud related interface, and we can wrap oss and gcs to implement this interface

I believe our use case is only around these scope:
- WriteAll (write all dags to respected bucket) 
- List (list all job from given bucket)
- Delete (delete dag file from bucket)
- Close

Reply:
as discussed in the devbox, current implementation will be kept for now. However, need to raise request to Ali team